### DISCUSSION TITLE
Subtle Liveware scenario where global scopes will be dropped potentially unexpectedly

### DISCUSSION THREAD

**Discussion by joshhanley** [livewire/livewire#7543] [7. Bugs (archived)]
### Livewire version

v3

### Laravel version

v10

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

This is an issue raised by @ianlandsman on Twitter regarding an issue with global scopes that may be applied to a model.

When Livewire rehydrates the model, the global scopes have been dropped, so it seems like a potential security concern, especially if those scopes are tenant scopes or something like that.

For example, if you use route model binding to find a user, that has a global scope applied like the following to restrict by domain:

```php
namespace App\Models\Scopes;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Scope;

class RestrictDomain implements Scope
{
    /**
     * Apply the scope to a given Eloquent query builder.
     */
    public function apply(Builder $builder, Model $model): void
    {
        $builder->where('email', 'LIKE', '%.com%');
    }
}
```

Then on the initial request, when Livewire uses route binding, the SQL query will look like this:

<img width="613" alt="image" src="https://github.com/livewire/livewire/assets/882837/3114d9a6-597e-4083-88b7-9634aa48f49c">

If you then store that user in a public property like:

```php
public User $user;

public function mount(User $user): void
{
    $this->user = $user;
}
```

When a Livewire request it triggered, and Livewire rehydrates the User model, the following SQL query is run:

<img width="509" alt="image" src="https://github.com/livewire/livewire/assets/882837/5c2985ed-070e-442b-8ad8-d9342b57d0ac">

As we can see there is no restrict domain scope in the query.

### Repository or Wirebox URL

https://github.com/ianlandsman/livewire-scopes

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joshhanley** [comment]
So looking into this, what controls the rehydration of the user model, is the `ModelSynth`.

Specifically this line, handles the rehydration of the model and making a database query to get the models attributes:

https://github.com/livewire/livewire/blob/40888d05443efd885022bc00c1201b28021deb8c/src/Features/SupportModels/ModelSynth.php#L57

What is happening here is Livewire is using the same process that Laravel uses in queues to restore a serialised model:

https://github.com/laravel/framework/blob/59f184efac7d9cc59a3ee06a6d881d87f00f9533/src/Illuminate/Queue/SerializesAndRestoresModelIdentifiers.php#L107-L109

In this scenario, both Livewire and Laravel require a model to have already been fetched from the database before it is serialised into the meta data required for the above code to work. The data that is stored is the model class and ID (or IDs if using an Eloquent Collection).

So the assumption is that any scopes/ security/ etc have already been applied to the model to retrieve it from the database in the first place.

That way when Livewire/ Laravel restore the model, they only need to perform the `whereIn` query passing along the ID/IDs.

We can see this happening in the `newQueryForRestoration()` method called on the model class above:

```php
public function newQueryForRestoration($ids)
{
    return $this->newQueryWithoutScopes()->whereKey($ids);
}
```

Looking at the method above, we can see that none of the global scopes are applied, due to the `newQueryWithoutScopes()`.

Now in a Laravel queue, this isn't an issue at all, as there is no scenario where a user can modify the model class and ID that Laravel has serialised, so it's safe to restore the models using the same information (and without any global scopes).

What about in a Livewire scenario though? The model class and ID are sent to the front end, so there is risk of a user being able to tamper with this.

In Livewire V3, Eloquent Model properties can't be `wire:model` or modified  in the front end at all. So the serialised model class and ID are stored inside Livewire's component snapshot, which is protected from tampering by Livewire's checksum. No other model attributes are sent to the front end.

This means, that the model class and ID can't be modified by the user/ front end either, otherwise a checksum exception will be thrown.

Based on this, storing Eloquent Models that contain scopes in public properties should be as secure as Laravel's queue restorations.

If anyone see's any issues with this, please let me know 🙂

Hope this helps!

**Comment by calebporzio** [comment]
Thanks Josh, those were my thoughts exactly. Once the model is set, it can't be changed externally to a model of another tenant, only from PHP code can that happen. Thanks for the writeup and and great explanation!

**Reply by joshhanley** [reply]
No worries! 🙂
